Improvements to CoreCLR Building, Debugging, and Testing Documentation (#75569)
dotnet/runtime: docs/workflow/testing/coreclr/testing.md

# Building and Running CoreCLR Tests

* [Requirements](#requirements)
* [Overview](#overview)
* [Building the Core_Root](#building-the-core_root)
* [Building the Tests](#building-the-tests)
  * [Building an Individual Test](#building-an-individual-test)
  * [Building a Test Directory](#building-a-test-directory)
  * [Building a Test Subtree](#building-a-test-subtree)
  * [Building C++/CLI Native Test Components Against the Live Ref Assemblies](#building-ccli-native-test-components-against-the-live-ref-assemblies)
  * [Test Priorities](#test-priorities)
* [Running the Tests](#running-the-tests)
  * [Running Individual Tests](#running-individual-tests)
    * [Tests Without a Main Method](#tests-without-a-main-method)
  * [PAL Tests (MacOS and Linux only)](#pal-tests-macos-and-linux-only)
    * [Building PAL Tests](#building-pal-tests)
    * [Running PAL Tests](#running-pal-tests)
* [Modifying Tests](#modifying-tests)
* [Investigating Test Failures](#investigating-test-failures)

This guide will walk you through building and running the CoreCLR tests. These are located within the `src/tests` subtree of the runtime repo.

## Requirements

> In order to build CoreCLR tests, you will need to have built the runtime and the libraries (i.e. _clr_ and _libs_ subsets). You can find more detailed instructions per platform in their dedicated docs:

Reviewer: ```suggestion
In order to build CoreCLR tests, you will need to have built the runtime and the libraries (that is, _clr_ and _libs_ subsets). You can find more detailed instructions per platform in their dedicated docs:
```

* `-env`: Path to a _.env_ file to specify environment variables to be set for the test. More info about _dotenv_ can be found in [their repo](https://github.com/motdotla/dotenv).
* -coreroot: The path to the Core_Root you wish to use. Note that this flag is mandatory unless you have the `CORE_ROOT` environment variable set. Then, you can omit it and the script will use that one.

If this list of parameters feels familiar, it's because it's virtually the same as the arguments that `corerun` receives. More info on `corerun` in its [how-to-use doc](/docs/workflow/testing/using-corerun-and-coreroot.md).

These scripts have a couple more hidden functionalities, which can be activated by setting their environment variables prior to running the script:

* Run with Crossgen2: `RunCrossGen2=1`
* Build and run as composite: `CompositeBuildMode=1`. Note that this one depends on `RunCrossGen2` being set.

Let's run one of the Intrinsics tests as an example:

On Windows:

```cmd
set CORE_ROOT=<repo_root>\artifacts\tests\coreclr\windows.<Arch>.<Configuration>\Tests\Core_Root
cd path\to\JIT\Intrinsics\MathRoundDouble_ro
.\MathRoundDouble_ro.cmd
```

On MacOS/Linux:

```bash

> export CORE_ROOT=<repo_root>/artifacts/tests/coreclr/<OS>.<Arch>.<Configuration>/Tests/Core_Root

Reviewer 2: You can also pass `coreroot=<repo_root>/artifacts/tests/coreclr/<OS>.<Arch>.<Configuration>/Tests/Core_Root` argument to the tests .sh. This is my preferred way. Similarly on Windows, but this time the `coreroot` is separated by space from the path, not by `=`.
